akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcProjectionImpl.scala

```diff
@@ -701,30 +702,33 @@ private[projection] object R2dbcProjectionImpl {
     }
 
     val validate = Flow[(Envelope, ProjectionContext)]
-      .mapAsync(1) {
+      .flatMapConcat {
         case (env, context) =>
-          offsetStore
-            .validate(env)
-            .flatMap {
-              case Accepted =>
-                accepted(env, context)
-              case Duplicate =>
-                Future.successful(None)
-              case RejectedSeqNr =>
-                replayIfPossible(env, observer(context)).flatMap {
-                  // if missing events were replayed immediately, then accept the rejected envelope for downstream processing
-                  case ReplayedOnRejection => accepted(env, context)
-                  case ReplayTriggered     => Future.successful(None)
-                  case NotReplayed         => Future.successful(None)
-                }
-              case RejectedBacktrackingSeqNr =>
-                replayIfPossible(env, observer(context)).flatMap {
-                  // if missing events were replayed immediately, then accept the rejected envelope for downstream processing
-                  case ReplayedOnRejection => accepted(env, context)
-                  case ReplayTriggered     => Future.successful(None)
-                  case NotReplayed         => throwRejectedEnvelope(sourceProvider, env)
-                }
-            }
+          // Note: while replays are a submaterialized source, they are in fact folded into a single source element
+          val futureSource: Future[Source[Option[(Envelope, ProjectionContext)], NotUsed]] =
+            offsetStore
+              .validate(env)
+              .map[Source[Option[(Envelope, ProjectionContext)], NotUsed]] {
+                case Accepted =>
+                  Source.future(accepted(env, context))
```

Comment:
would it be optimized if `accepted` could be a source? I think in normal cases it will already be completed futures, and we might be able to do some optimization of already completed `Future(Source.single)`? we have some optimization for Source.singe in streams?